pkg/operator/ceph/object/spec.go

```diff
@@ -0,0 +1,219 @@
+	container := v1.Container{
+		Name:  opspec.ConfigInitContainerName,
+		Image: k8sutil.MakeRookImage(version),
+		Args: []string{
+			"ceph",
+			"rgw",
+			fmt.Sprintf("--config-dir=%s", k8sutil.DataDir),
+			fmt.Sprintf("--rgw-name=%s", store.Name),
+			fmt.Sprintf("--rgw-port=%d", store.Spec.Gateway.Port),
+			fmt.Sprintf("--rgw-secure-port=%d", store.Spec.Gateway.SecurePort),
+		},
+		VolumeMounts: opspec.RookVolumeMounts(),
+		Env: []v1.EnvVar{
+			{Name: "ROOK_RGW_KEYRING", ValueFrom: &v1.EnvVarSource{SecretKeyRef: &v1.SecretKeySelector{LocalObjectReference: v1.LocalObjectReference{Name: instanceName(store)}, Key: keyringName}}},
+			k8sutil.PodIPEnvVar(k8sutil.PrivateIPEnvVar),
+			k8sutil.PodIPEnvVar(k8sutil.PublicIPEnvVar),
+			opmon.ClusterNameEnvVar(store.Namespace),
+			opmon.EndpointEnvVar(),
+			opmon.SecretEnvVar(),
+			k8sutil.ConfigOverrideEnvVar(),
+		},
+		Resources: store.Spec.Gateway.Resources,
+	}
+
+	if store.Spec.Gateway.SSLCertificateRef != "" {
+		// Add a volume mount for the ssl certificate
+		mount := v1.VolumeMount{Name: certVolumeName, MountPath: certMountPath, ReadOnly: true}
+		container.VolumeMounts = append(container.VolumeMounts, mount)
+
+		// Pass the flag for using the ssl cert
+		path := path.Join(certMountPath, certFilename)
+		container.Args = append(container.Args, fmt.Sprintf("--rgw-cert=%s", path))
+	}
+
+	return container
+}
+
+func makeDaemonContainer(store cephv1beta1.ObjectStore, version string) v1.Container {
+
+	// start the rgw daemon in the foreground
+	container := v1.Container{
+		Name:  "rgw",
+		Image: k8sutil.MakeRookImage(version),
+		Command: []string{
+			"radosgw",
+		},
+		Args: []string{
+			"--foreground",
+			"--name=client.radosgw.gateway",
+			fmt.Sprintf("--rgw-mime-types-file=%s", rgwdaemon.GetMimeTypesPath(k8sutil.DataDir)),
+		},
+		VolumeMounts: opspec.CephVolumeMounts(),
+		Env:          k8sutil.ClusterDaemonEnvVars(),
+		Resources:    store.Spec.Gateway.Resources,
+	}
+
+	return container
+}
+
+func startService(context *clusterd.Context, store cephv1beta1.ObjectStore, hostNetwork bool, ownerRefs []metav1.OwnerReference) (string, error) {
+	labels := getLabels(store)
+	svc := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      instanceName(store),
+			Namespace: store.Namespace,
+			Labels:    labels,
+		},
+		Spec: v1.ServiceSpec{
+			Selector: labels,
+		},
+	}
+	k8sutil.SetOwnerRefs(context.Clientset, store.Namespace, &svc.ObjectMeta, ownerRefs)
+	if hostNetwork {
+		svc.Spec.ClusterIP = v1.ClusterIPNone
+	}
+
+	addPort(svc, "http", store.Spec.Gateway.Port)
+	addPort(svc, "https", store.Spec.Gateway.SecurePort)
+
+	svc, err := context.Clientset.CoreV1().Services(store.Namespace).Create(svc)
+	if err != nil {
+		if !errors.IsAlreadyExists(err) {
+			return "", fmt.Errorf("failed to create rgw service. %+v", err)
+		}
+		logger.Infof("Gateway service already running")
+		return "", nil
```

Comment:
Is there no need to return the Service IP when it already exist?

Reply:
This was actually just moved from another file so I didn't change anything here. It must have been working like this, but I'll double check.

Reply:
There is an extremely small change this code path would ever be hit since the operator will skip the object store before this if the object store already exists, but I'll retrieve the service in this case just to be sure.